dipy/segment/metricspeed.pyx

```diff
@@ -0,0 +1,372 @@
+        cdef double dist_flipped = AveragePointwiseEuclideanMetric.c_dist(self, features1, features2[::-1])
+        return min(dist_direct, dist_flipped)
+
+
+cpdef distance_matrix(Metric metric, data1, data2=None):
+    """ Computes the distance matrix between two lists of sequential data.
+
+    The distance matrix is obtained by computing the pairwise distance of all
+    tuples spawn by the Cartesian product of `data1` with `data2`. If `data2`
+    is not provided, the Cartesian product of `data1` with itself is used
+    instead. A sequence of N-dimensional points is represented as a 2D array with
+    shape (nb_points, nb_dimensions).
+
+    Parameters
+    ----------
+    metric : `Metric` object
+        Tells how to compute the distance between two sequential data.
+    data1 : list of 2D array
+        List of sequences of N-dimensional points.
+    data2 : list of 2D array
+        Llist of sequences of N-dimensional points.
+
+    Returns
+    -------
+    2D array (double)
+        Distance matrix.
+    """
+    cdef int i, j
+    if data2 is None:
+        data2 = data1
+
+    shape = metric.feature.infer_shape(data1[0].astype(np.float32))
+    distance_matrix = np.zeros((len(data1), len(data2)), dtype=np.float64)
+    cdef:
+        Data2D features1 = np.empty(shape, np.float32)
+        Data2D features2 = np.empty(shape, np.float32)
+
+    for i in range(len(data1)):
+        datum1 = data1[i] if data1[i].flags.writeable and data1[i].dtype is np.float32 else data1[i].astype(np.float32)
+        metric.feature.c_extract(datum1, features1)
+        for j in range(len(data2)):
+            datum2 = data2[j] if data2[j].flags.writeable and data2[j].dtype is np.float32 else data2[j].astype(np.float32)
+            metric.feature.c_extract(datum2, features2)
+            distance_matrix[i, j] = metric.c_dist(features1, features2)
+
+    return distance_matrix
+
+
+cpdef double dist(Metric metric, datum1, datum2) except -1.0:
+    """ Computes distance between two sequential data.
+
+    A sequence of N-dimensional points is represented as a 2D array with
+    shape (nb_points, nb_dimensions).
+
+    Parameters
+    ----------
+    metric : `Metric` object
+        Tells how to compute the distance between `datum1` and `datum2`.
+    datum1 : 2D array
+        Sequence of N-dimensional points.
+    datum2 : 2D array
+        Sequence of N-dimensional points.
+
+    Returns
+    -------
+    double
+        Distance between two data points.
+    """
+    shape1 = metric.feature.infer_shape(datum1)
+    shape2 = metric.feature.infer_shape(datum2)
+
+    if not metric.are_compatible(shape1, shape2):
+        raise ValueError("Data features' shapes must be compatible!")
+
+    datum1 = datum1 if datum1.flags.writeable and datum1.dtype is np.float32 else datum1.astype(np.float32)
+    datum2 = datum2 if datum2.flags.writeable and datum2.dtype is np.float32 else datum2.astype(np.float32)
```

Comment:
I see some magic here. Do you know if the writeable is now fixed in the latest Cython version? Or this is not a bug?